packages/kit/src/core/sync/write_env.js

```diff
@@ -0,0 +1,76 @@
+import path from 'path';
+import { get_env } from '../../vite/utils.js';
+import { write_if_changed, reserved, valid_identifier } from './utils.js';
+
+const autogen_comment = '// this file is generated — do not edit it\n';
+
+/**
+ * Writes the existing environment variables in process.env to
+ * $env/static/private and $env/static/public
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string} mode
+ * The Vite mode.
+ */
+export function write_env(config, mode) {
+	const env = get_env(mode, config.env.publicPrefix);
+
+	// TODO when testing src, `$app` points at `src/runtime/app`... will
+	// probably need to fiddle with aliases
+	write_if_changed(
+		path.join(config.outDir, 'runtime/env/static/public.js'),
+		create_module('$env/static/public', env.public)
+	);
+
+	write_if_changed(
+		path.join(config.outDir, 'runtime/env/static/private.js'),
+		create_module('$env/static/private', env.private)
+	);
+
+	write_if_changed(
+		path.join(config.outDir, 'ambient.d.ts'),
+		autogen_comment +
+			create_types('$env/static/public', env.public) +
+			'\n\n' +
+			create_types('$env/static/private', env.private)
+	);
+}
+
+/**
+ * @param {string} id
+ * @param {Record<string, string>} env
+ * @returns {string}
+ */
+function create_module(id, env) {
+	const declarations = Object.entries(env)
+		.filter(([k]) => valid_identifier.test(k))
+		.map(([k, v]) => {
+			const comment = `/** @type {import('${id}'}').${k}} */`;
+
+			if (reserved.has(k)) {
+				let i = 1;
+				while (k + i in env) i++;
+				const name = k + 1;
+
+				return `${comment}\nconst ${name} = ${JSON.stringify(v)}; export { ${name} as ${k} }`;
+			}
```

Comment:
I'd still like a resolution on this (see discussion starting [here](https://discordapp.com/channels/457912077277855764/997637728608985169/1001258747508830308)).

I'd really prefer to log a warning and ignore reserved keywords.

Reply:
I mean, we're _already_ silently filtering out invalid JS identifiers. Seems like we should just silently filter out reserved JS keywords as well and make sure it's notated in the documentation. I'd be fine with that. Though the logging is probably preferable?

Reply:
Added some warnings for this. They'll be hidden by the `vite dev` welcome message, but rather than making our workaround more all-encompassing (i started down this road, it gets a little messy) i opened an issue on vite: https://github.com/vitejs/vite/issues/9378